More detailed error when attempting re-use of GPDir

Related to the issues/errors outlined in https://github.com/els0r/goProbe/pull/141: if the directory isn't open, make sure to communicate it properly.

For the record: Crash trace on commit 865a64d:
```
./goQuery --logging.level=debug -d /tmp/db_out2 -f '28.6.18 23:00' -l '28.6.18 23:30:00' list eth0
ts=2023-08-05T12:30:56.305+02:00 level=debug msg="handling first directory" version=872abf81 dir=/tmp/db_out2/eth0/2018/06/1530144000
ts=2023-08-05T12:30:56.305+02:00 level=debug msg="subtracting surplus blocks before tfirst" version=872abf81 blocks=288 surplus=252
ts=2023-08-05T12:30:56.307+02:00 level=debug msg="handling last directory" version=872abf81 dir=/tmp/db_out2/eth0/2018/06/1530144000
ts=2023-08-05T12:30:56.308+02:00 level=debug msg="opening last directory" version=872abf81
ts=2023-08-05T12:30:56.308+02:00 level=debug msg="checking last timestamp" version=872abf81 last=1530230129
ts=2023-08-05T12:30:56.308+02:00 level=debug msg="subtracting surplus blocks after tlast" version=872abf81 last=1530230129 blocks=288 surplus=29 offset=259
panic: runtime error: index out of range [259] with length 0

goroutine 1 [running]:
github.com/els0r/goProbe/pkg/goDB/storage/gpfile.(*GPFile).ReadBlockAtIndex(0x4?, 0xc00030c1f0?)
	/home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/goDB/storage/gpfile/gpfile.go:136 +0x60d
github.com/els0r/goProbe/pkg/goDB/storage/gpfile.(*GPDir).ReadBlockAtIndex(0xc0005c01e0, 0x4, 0x0?)
	/home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/goDB/storage/gpfile/gpdir.go:212 +0x4d
github.com/els0r/goProbe/pkg/goDB.(*DBWorkManager).readMetadataAndEvaluate(0xc00031e570?, 0xc0005c01e0, 0xc0000b8bd0, {0xc00039a878, 0x1d, 0x2000?}, 0x103, 0xc000110e00, 0xbe8858)
	/home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/goDB/DBWorkManager.go:434 +0xbc5
github.com/els0r/goProbe/pkg/goDB.(*DBWorkManager).ReadMetadata(0xc0001bce40, 0x5b354c50, 0x5b355358)
	/home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/goDB/DBWorkManager.go:357 +0xa51
github.com/els0r/goProbe/cmd/goQuery/cmd.listInterfaces({0x7ffefaa49e00, 0xc}, {0xc000110d80, 0x1, 0x0?})
	/home/fako/Develop/go/src/github.com/els0r/goProbe/cmd/goQuery/cmd/list.go:103 +0x3cd
github.com/els0r/goProbe/cmd/goQuery/cmd.listInterfacesEntrypoint(0x10908a0?, {0xc000110d80, 0x1, 0x8})
	/home/fako/Develop/go/src/github.com/els0r/goProbe/cmd/goQuery/cmd/list.go:50 +0x4f
github.com/spf13/cobra.(*Command).execute(0x10908a0, {0xc0000a8460, 0x8, 0xe})
	/home/fako/Develop/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0x1090b80)
	/home/fako/Develop/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/home/fako/Develop/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/els0r/goProbe/cmd/goQuery/cmd.Execute()
	/home/fako/Develop/go/src/github.com/els0r/goProbe/cmd/goQuery/cmd/root.go:55 +0x32
main.main()
	/home/fako/Develop/go/src/github.com/els0r/goProbe/cmd/goQuery/main.go:6 +0x17
```

--- a/pkg/goDB/storage/gpfile/gpdir.go
+++ b/pkg/goDB/storage/gpfile/gpdir.go
@@ -36,6 +36,9 @@ var (
 
 	// ErrInputSizeTooSmall is thrown if the input size for desiralization is too small to be valid
 	ErrInputSizeTooSmall = errors.New("input size too small to be a GPDir metadata header")
+
+	// ErrDirNotOpen denotes that a GPDir is not (yet) open or has been closed
+	ErrDirNotOpen = errors.New("GPDir not open, call Open() first")
 )
 
 // TrafficMetadata denotes a serializable set of metadata information about traffic stats
@@ -120,6 +123,7 @@ type GPDir struct {
 	accessMode  int         // Access mode (also forwarded to all GPFiles)
 	permissions os.FileMode // Permissions (also forwarded to all GPFiles)
 
+	isOpen bool
 	*Metadata
 }
 
@@ -186,6 +190,7 @@ func (d *GPDir) Open(options ...Option) error {
 		}
 	}
 
+	d.isOpen = true
 	return nil
 }
 
@@ -201,6 +206,10 @@ func (d *GPDir) NumIPv6EntriesAtIndex(blockIdx int) uint64 {
 
 // ReadBlockAtIndex returns the block for a specified block index from the underlying GPFile
 func (d *GPDir) ReadBlockAtIndex(colIdx types.ColumnIndex, blockIdx int) ([]byte, error) {
+
+	if !d.isOpen {
+		return nil, ErrDirNotOpen
+	}
 
 	// Load column if required
 	_, err := d.Column(colIdx)
@@ -423,6 +432,10 @@ func (d *GPDir) NBlocks() int {
 // Close closes all underlying open GPFiles and cleans up resources
 func (d *GPDir) Close() error {
 
+	defer func() {
+		d.isOpen = false
+	}()
+
 	// Close all open GPFiles
 	var errs []error
 	for i := 0; i < int(types.ColIdxCount); i++ {
@@ -459,6 +472,11 @@ func (d *GPDir) Close() error {
 
 // column returns the underlying GPFile for a specified column (lazy-access)
 func (d *GPDir) Column(colIdx types.ColumnIndex) (*GPFile, error) {
+
+	if !d.isOpen {
+		return nil, ErrDirNotOpen
+	}
+
 	if d.gpFiles[colIdx] == nil {
 		var err error
 		if d.gpFiles[colIdx], err = New(filepath.Join(d.Path(), types.ColumnFileNames[colIdx]+FileSuffix), d.BlockMetadata[colIdx], d.accessMode, d.options...); err != nil {

--- a/pkg/goDB/storage/gpfile/gpfile_test.go
+++ b/pkg/goDB/storage/gpfile/gpfile_test.go
@@ -316,7 +316,7 @@ func TestMetadataRoundTrip(t *testing.T) {
 	require.Equal(t, sumDrops, int(testDir.Metadata.Traffic.NumDrops), "mismatched number of total packet drops vs. computed")
 }
 
-func TestBrokenWrite(t *testing.T) {
+func TestBrokenAccess(t *testing.T) {
 
 	require.Nil(t, os.RemoveAll("/tmp/test_db"))
 
@@ -344,7 +344,20 @@ func TestBrokenWrite(t *testing.T) {
 	require.Nil(t, testDir.Open(), "error opening test dir for reading")
 	require.Equal(t, expectedOffset, testDir.BlockMetadata[0].CurrentOffset)
 	require.Equal(t, testDir.BlockMetadata[0].NBlocks(), 2)
+	for i := types.ColumnIndex(0); i < types.ColIdxCount; i++ {
+		_, err := testDir.ReadBlockAtIndex(i, 0)
+		require.Nil(t, err)
+	}
 	require.Nil(t, testDir.Close(), "error closing test dir")
+
+	// Attempt to read from closed GPDir (should return an error)
+	t.Run("not open", func(t *testing.T) {
+		for i := types.ColumnIndex(0); i < types.ColIdxCount; i++ {
+			data, err := testDir.ReadBlockAtIndex(i, 0)
+			require.Nil(t, data)
+			require.ErrorIs(t, err, ErrDirNotOpen)
+		}
+	})
 
 	// Append another two blocks and write normally
 	testDir = NewDir("/tmp/test_db", 1000, ModeWrite)

